tripal/tripal
# publication title and pyear are nullable, or so says chado

# Minor Bug Fix

### Closes #2015

### Tripal Version: 4.x

## Description
The `pyear` field for the publication content type is set in the yaml as a required field. Likewise for `title`.
This matches Tripal 3 behavior, however chado defines these fields as nullable.
This is an issue for publish, where we select one required field for finding existing entities.
If a publication is imported without a `pyear` or `title` value, or it is entered into the database directly, then if the pub content type is published, duplicate records can result. This happens notably with the null publication.

An update hook is provided for existing sites.

## Testing?
1. Build on 4.x branch so we can test the update hook
2. Switch to this branch `git checkout tv4g1-issue2015-publication-pyear`
3. Run update hook `drush updatedb`
```
 -------------- ----------- --------------- ---------------------------------- 
  Module         Update ID   Type            Description                       
 -------------- ----------- --------------- ---------------------------------- 
  tripal_chado   10409       hook_update_n   10409 - Sets publication 'title'  
                                             and 'pyear' fields as             
                                             non-required, to be consistent    
                                             with chado. PR 2029               
 -------------- ----------- --------------- ---------------------------------- 


 ┌ Do you wish to run the specified pending updates? ───────────┐
 │ Yes                                                          │
 └──────────────────────────────────────────────────────────────┘

>  [notice] Update started: tripal_chado_update_10409
>  [notice] Update completed: tripal_chado_update_10409
>  [notice] Message: Changed field tripal_entity.pub.pub_title to be not required
> 
>  [notice] Message: Changed field tripal_entity.pub.pub_pyear to be not required
> 
 [success] Finished performing updates.
```
4. Publish the publication content type.
5. Publish publications a second time. Should have nothing published.
On 4.x you would have seen: `Publish completed. Published 1 new entities, and added 4 field values.`
On this branch you should see: `There are no records to publish for this content type`
6. Count how many null publications there are. Should only be one.
![PR2029 1](https://github.com/user-attachments/assets/79cdb0c3-419b-4fbf-968d-f2953dddbd3c)
7. Enter a new publication through UI and verify that title and publication year do not have an asterisk, i.e. they are no longer required.
